### DISCUSSION TITLE
Pico as usb host

### DISCUSSION THREAD

**Discussion by emojogurt** [earlephilhower/arduino-pico#224] [Q&A] [answered]
Hello, 

I was wondering, how could one approach using usb host functionality on pico using this?

Would that be even possible?

What I tried so far was to use the example from tinyusb library that I already managed to run from 
'pure' pico-sdk. Of course, this did not work directly due to tusb_config.h (I might be confused as to the filename now, sorry if that is the case). I modified it to bring it in line with what was needed, but the usb device usage started popping up here and there in the files for the board.

Naturally, the first thing that comes to mind is to go over those places and comment them out/ delete the files, but...

Now, if I understand correctly, the board functioning as usb device is necessary to use e.g. Serial functionality, but if one resigned oneself to *not* using it, is this a feasible option to just remove all of those usages, or will it cause everything to come crashing down and prevent me from uploading / building the project?

**Comment by earlephilhower** [comment, marked as answer]
It may be possible to do this, but you'll need to not use `Serial` as you stated and also add a define `DISABLE_USB_SERIAL` or hack out
https://github.com/earlephilhower/arduino-pico/blob/35c974d4c0fc185270b9ca5f1dbb8650ba702574/cores/rp2040/main.cpp#L70-L73

You'd then have to handle all the USB setup yourself, as there is no host mode code in the core yet.

Unless you need the Arduino libraries, though, I'd probably suggest just using the SDK for host mode apps at this point.

Of course, a PR adding support for USB host mode would definitely be welcome! :)

**Reply by emojogurt** [reply]
First of all, thank you for the quick answer!

What you're saying is very encouraging, I'll be definitely giving this a try in the coming days.

As to the PR, it would definitely be great to contribute, but my coding skills might prove insufficient here...

I'll definitely share if I get anywhere, though.

Thanks again!

**Reply by emojogurt** [reply]
Hello, again. I've made some progress in that area... but it might have been going in a wrong direction, as I didn't really manage to have this functioning. Still, what I did might help someone smarter than me figure this out, so I'm putting my story here anyway.

What I have is a (almost) minimal example of USB host for a HID device (Keyboard in my case), that was built by cutting out everything unneeded from the (already basic) HID example from TinyUSB.

Basically it does only one thing: blinks a led when you press a button on the keyboard. For two of the keys (a and p, no special significance), there are special blink sequences just to make sure the scancodes are read correctly - this is only meant to be a POC of reading the keystrokes. 

(background: I need this, since my Arduino project that uses a keyboard can only support PS/2 due to Arduino limitations, and as it happens the keyboard I wanted to use can't deal with that, so I hoped to resolve this with pico in Arduino IDE - this way only the keyboard part would have to be re-written. Added benefit would be that the project will be more up-to-date with the state of technology today, I don't even remember last time I plugged a PS/2 keyboard to a computer, but I digress...)

main.c is:
(see attachments for full source):
int main(void)
{
  stdio_init_all();
  board_init();
  tusb_init();
  while (1)
  {
    tuh_task();
  }
  return 0;
}

So, first step was to put this into the main sketch for arduino and try to figure out necessary includes.

1. I added:
#include "tusb.h"

at the top of the sketch. Unfortunately, as I already mentioned in my initial post, that started giving me issues (not recognising the tuh_task and asking whether I meant tud_task) due to default tusb_config.h, so...

2. I added the config file (see attachment) and the define (as advised). Please note, that the config must have the specific #ifndef at the top, otherwise the settings will be overwritten by the default config from the board files.

#define DISABLE_USB_SERIAL
#include "tusb_config.h"

This did not resolve the issue, as the error (though the message was slightly different) still informed me of:
undefined reference to tuh_task. 
So...

3. After some searching, I realised that not only the headers, but also the .c files must be included in the sketch (??? I don't understand why, but it doesn't really work otherwise...). I took them from the Makefile generated by the cmake for the example in pico-sdk, and those were:

#include "<path to board files in my system>/1.8.8/pico-sdk/lib/tinyusb/src/host/usbh.h"
#include "<path to board files in my system>/1.8.8/pico-sdk/lib/tinyusb/src/host/usbh.c"
Note, that there are more, but those two included the tuh_task declaration ond definition.

(I had to include full paths, because I don't know how do I tell the compiler to search for them in another...)

This indeed resolved the issue of 'undefined reference to tuh_task', but, sadly, I started getting errors about invalid conversions in some of the files. Since I knew the example compiled OK in the pico-sdk, I just suppressed them by adding -fpermissive to the platform.txt in the board files.

build.extra_flags=-fpermissive

those errors were indeed replaced by warnings, but I started getting another one:

/hardware/rp2040/1.8.8//pico-sdk/lib/tinyusb/src/osal/osal_none.h:131:3: error: designator order for field 'tu_fifo_t::non_used_index_space' does not match declaration order

Since at this stage I was way out of my depth, I googled this, and a comment on reddit gave me a clue as to what this might be:
https://www.reddit.com/r/cpp/comments/70poxv/should_c_have_tagged_initialization/dn50ndf?utm_source=share&utm_medium=web2x&context=3

Apparently, some code in this file (or, rather, the code that calls the macro defined there) is valid in C, but not in C++ 
(frankly, I'm not sure if that is really the case).

This is the stage at which I'm now, and I think this is as far as I can get.

Code working in the tinyusb examples:
[CMakeLists.txt](https://github.com/earlephilhower/arduino-pico/files/6721530/CMakeLists.txt)
src directory:
[hid_app.c.txt](https://github.com/earlephilhower/arduino-pico/files/6721474/hid_app.c.txt)
[main.c.txt](https://github.com/earlephilhower/arduino-pico/files/6721475/main.c.txt)
[tusb_config.h.txt](https://github.com/earlephilhower/arduino-pico/files/6721476/tusb_config.h.txt)

I'm going to try my luck at porting the project to pico-sdk now...

**Reply by earlephilhower** [reply]
FWIW you can just copy `*.c` files into the sketch directory and the IDE will use the C compiler and not C++ one on them.  Not sure it makes what you're doing any easier, mind you!

**Comment by karmanyaahm** [comment]
I was running into the same problem with `tuh_task` and then I copied all of tinyusb (v12, which is what pico-sdk has) sources into my platformio project's src folder (as suggested by @earlephilhower above) and it worked.
@earlephilhower thx for making this toolchain, it's really helpful for my project
